cmd/xl-storage.go

```diff
@@ -1180,7 +1180,10 @@ func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions
 			errs[i] = ctx.Err()
 			continue
 		}
-		w := xioutil.NewDeadlineWorker(globalDriveConfig.GetMaxTimeout())
+		// Scale the deadline with the number of versions for this object since an object
+		// can contain thousands of versions (though not recommended) and removing a version
+		// can involve moving data-dir to the trash.
+		w := xioutil.NewDeadlineWorker(globalDriveConfig.GetMaxTimeout() + time.Duration(len(fiv.Versions))*time.Second)
```

Comment:
this value is too much this is not how it happens internally this needs to be fixed to include only to "specific" object names.

Because deleting X versions of an object is mostly atomic i.e it purges at instant

Reply:
the purge is not instant if versions have data-dir. deleteVersions() will move data-dir to the trash folder and that sometimes takes a lot of time, more than the configured deadline (30 seconds is the default one). This means if one object has 10k versions, then this function will fail with a deadline. That is why I wanted to scale it depending on the number of versions of the object.

Reply:
yeah but that is literally forever, any drive that takes that long is dead already, there is no point performing this operation - we timeout and let other drives complete. 

This is not the place the optimize TBH.

Reply:
we have no reason to hold the the caller this long, a drive takes 30secs to rename you need to throw that drive. 

They are free to configure higher MaxTimeout per drive but this not correct we adding arbitrary number of seconds.

Reply:
This use case is not a single rename, it is thousands of rename(), users will also have to take an arbitrary number.

It looks to me that a more comprehensive fix is to do a deadline per function inside deleteVersions(), or maybe a better option is to add new API to deadline worker to delay triggering the deadline

Reply:
that's what we are doing here

```
// DeleteVersions deletes slice of versions, it can be same object                                                                                            
// or multiple objects.                                                                                                                                       
func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions []FileInfoVersions, opts DeleteOptions) []error {
        errs := make([]error, len(versions))

        for i, fiv := range versions {
                if contextCanceled(ctx) {
                        errs[i] = ctx.Err()
                        continue
                }
                w := xioutil.NewDeadlineWorker(globalDriveConfig.GetMaxTimeout())
                if err := w.Run(func() error { return s.deleteVersions(ctx, volume, fiv.Name, fiv.Versions...) }); err != nil {
                        errs[i] = err
                }
                diskHealthCheckOK(ctx, errs[i])
        }

        return errs
}
```

Reply:
each object gets a max 30sec to expire,there is no reason to have top level deadline here.